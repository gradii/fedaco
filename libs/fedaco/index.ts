/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license
 */

export * from './src/interface/connection-resolver-interface';
export * from './src/base-grammar';
export * from './src/define/collection';
export * from './src/connection/sql-server-connection';
export * from './src/connection/mysql-connection';
export * from './src/connection/sqlite-connection';
export * from './src/connection/postgres-connection';
export * from './src/fedaco/model-helper-global';
export * from './src/fedaco/fedaco-builder';
export * from './src/fedaco/fedaco-event-manage';
export * from './src/fedaco/model-meta-data';
export * from './src/fedaco/base-model';
export * from './src/fedaco/scope';
export * from './src/fedaco/model-resolver';
export * from './src/fedaco/relations/pivot';
export * from './src/fedaco/relations/pivot-helper';
export * from './src/fedaco/relations/has-one-or-many';
export * from './src/fedaco/relations/has-many-through';
export * from './src/fedaco/relations/morph-many';
export * from './src/fedaco/relations/has-many';
export * from './src/fedaco/relations/belongs-to-many';
export * from './src/fedaco/relations/morph-one-or-many';
export * from './src/fedaco/relations/morph-one';
export * from './src/fedaco/relations/has-one';
export * from './src/fedaco/relations/belongs-to';
export * from './src/fedaco/relations/has-one-through';
export * from './src/fedaco/relations/morph-pivot';
export * from './src/fedaco/relations/morph-to-many';
export * from './src/fedaco/relations/relation';
export * from './src/fedaco/relations/concerns/interacts-with-pivot-table';
export * from './src/fedaco/relations/concerns/can-be-one-of-many';
export * from './src/fedaco/relations/concerns/compares-related-models';
export * from './src/fedaco/relations/concerns/supports-default-models';
export * from './src/fedaco/relations/concerns/as-pivot';
export * from './src/fedaco/relations/concerns/interacts-with-dictionary';
export * from './src/fedaco/relations/morph-to';
export * from './src/fedaco/mixins/soft-deletes';
export * from './src/fedaco/mixins/has-global-scopes';
export * from './src/fedaco/mixins/guards-attributes';
export * from './src/fedaco/mixins/hides-attributes';
export * from './src/fedaco/mixins/has-timestamps';
export * from './src/fedaco/mixins/queries-relationships';
export * from './src/fedaco/mixins/has-relationships';
export * from './src/fedaco/mixins/forward-call-to-query-builder';
export * from './src/fedaco/mixins/has-events';
export * from './src/fedaco/mixins/has-attributes';
export * from './src/fedaco/model-helper';
export * from './src/fedaco/model';
export * from './src/fedaco/fedaco-manage';
export * from './src/fedaco/resolve-connection';
export * from './src/fedaco/encrypter';
export * from './src/fedaco/scopes/soft-deleting-scope';
// export * from './src/dbal/column';
// export * from './src/dbal/table';
// export * from './src/dbal/table-diff';
export * from './src/annotation/column';
export * from './src/annotation/additional-processing';
// export * from './src/annotation/scope';
export * from './src/annotation/enum-relation';
export * from './src/annotation/column/column';
export * from './src/annotation/column/binary.column';
export * from './src/annotation/column/updated-at.column';
export * from './src/annotation/column/primary-generated.column';
export * from './src/annotation/column/integer.column';
export * from './src/annotation/column/array.column';
export * from './src/annotation/column/float.column';
export * from './src/annotation/column/datetime.column';
export * from './src/annotation/column/decimal.column';
export * from './src/annotation/column/created-at.column';
export * from './src/annotation/column/timestamp.column';
export * from './src/annotation/column/primary.column';
export * from './src/annotation/column/object.column';
export * from './src/annotation/column/deleted-at.column';
export * from './src/annotation/column/date.column';
export * from './src/annotation/column/text.column';
export * from './src/annotation/column/boolean.column';
export * from './src/annotation/column/json.column';
export * from './src/annotation/column/currency.column';
export * from './src/annotation/annotation.interface';
export * from './src/annotation/table/table';
export * from './src/annotation/boot';
export * from './src/annotation/relation-using.relation';
export * from './src/annotation/featured-column/auto-number.column';
// export * from './src/annotation/featured-column/paragraph.column';
export * from './src/annotation/relation-column';
export * from './src/annotation/relation-column/has-many-through.relation-column';
export * from './src/annotation/relation-column/morph-one.relation-column';
export * from './src/annotation/relation-column/morph-many.relation-column';
export * from './src/annotation/relation-column/morph-to-many.relation-column';
export * from './src/annotation/relation-column/belongs-to.relation-column';
export * from './src/annotation/relation-column/belongs-to-many.relation-column';
export * from './src/annotation/relation-column/has-many.relation-column';
export * from './src/annotation/relation-column/morphed-by-many.relation-column';
export * from './src/annotation/relation-column/morph-eager-to.relation-column';
export * from './src/annotation/relation-column/has-one.relation-column';
export * from './src/annotation/relation-column/morph-to.relation-column';
export * from './src/annotation/relation-column/has-one-through.relation-column';
export * from './src/connection';
export * from './src/schema/foreign-id-column-definition';
export * from './src/schema/postgres-schema-state';
export * from './src/schema/grammar/mysql-schema-grammar';
export * from './src/schema/grammar/postgres-schema-grammar';
export * from './src/schema/grammar/sqlite-schema-grammar';
export * from './src/schema/grammar/sql-server-schema-grammar';
// export * from './src/schema/grammar/rename-column';
export * from './src/schema/grammar/schema-grammar';
// export * from './src/schema/grammar/change-column';
export * from './src/schema/schema-state';
export * from './src/schema/sqlite-schema-state';
export * from './src/schema/column-definition';
export * from './src/schema/schema-builder';
export * from './src/schema/mysql-schema-state';
export * from './src/schema/foreign-key-definition';
export * from './src/schema/blueprint';
export * from './src/schema/builder/sql-server-schema-builder';
export * from './src/schema/builder/sqlite-schema-builder';
export * from './src/schema/builder/postgres-schema-builder';
export * from './src/schema/builder/mysql-schema-builder';
export * from './src/connector/sql-server-connector';
// export * from './src/connector/mariadb/mariadb-connector';
// export * from './src/connector/mariadb/mariadb-wrapped-connection';
// export * from './src/connector/mariadb/mariadb-wrapped-stmt';
export * from './src/connector/wrapped-connection';
export * from './src/connector/connector';
export * from './src/connector/sqlite/sqlite-wrapped-connection';
export * from './src/connector/sqlite/sqlite-connector';
export * from './src/connector/sqlite/sqlite-wrapped-stmt';
export * from './src/connector/mysql/mysql-wrapped-connection';
export * from './src/connector/mysql/mysql-connector';
export * from './src/connector/mysql/mysql-wrapped-stmt';
export * from './src/connector/wrapped-stmt';
export * from './src/connector/postgres-connector';
export * from './src/connector/connector-interface';
export * from './src/connector/connection-factory';
export * from './src/database-manager';
export * from './src/events/query-executed';
export * from './src/events/connection-event';
export * from './src/events/schema-loaded';
export * from './src/events/transaction-rolled-back';
export * from './src/events/schema-dumped';
export * from './src/events/statement-prepared';
export * from './src/events/no-pending-migrations';
export * from './src/events/transaction-committed';
export * from './src/events/models-pruned';
export * from './src/events/transaction-beginning';
export * from './src/query/sql-node';
export * from './src/query/expr/functions/function-node';
export * from './src/query/expr/functions/abs-function';
export * from './src/query/parser/sql-parser';
export * from './src/query/parser/helper';
// export * from './src/query/parser/_sql-parser-ast';
export * from './src/query/parser/sql-lexer';
export * from './src/query/parser/keywords';
export * from './src/query/parser/ascii-chars';
export * from './src/query/sql-visitor';
export * from './src/query/ast/table-name';
export * from './src/query/ast/path-expression';
export * from './src/query/ast/join-expression';
export * from './src/query/ast/select-insert-source';
export * from './src/query/ast/identifier';
export * from './src/query/ast/column-reference-expression';
export * from './src/query/ast/order-by-element';
export * from './src/query/ast/select-clause';
export * from './src/query/ast/join-on-expression';
export * from './src/query/ast/having-clause';
export * from './src/query/ast/delete-specification';
export * from './src/query/ast/join-clause';
export * from './src/query/ast/index-by';
export * from './src/query/ast/group-by-clause';
export * from './src/query/ast/range-variable-declaration';
export * from './src/query/ast/offset-clause';
export * from './src/query/ast/binary-query-expression';
export * from './src/query/ast/values-insert-source';
export * from './src/query/ast/where-clause';
export * from './src/query/ast/binary-union-query-expression';
export * from './src/query/ast/assignment-set-clause';
export * from './src/query/ast/query-expression';
export * from './src/query/ast/fragment/order/reject-order-element-expression';
export * from './src/query/ast/fragment/union-fragment';
export * from './src/query/ast/fragment/aggregate-fragment';
export * from './src/query/ast/fragment/nested-expression';
export * from './src/query/ast/fragment/raw-fragment';
export * from './src/query/ast/fragment/expression/nested-predicate-expression';
export * from './src/query/ast/fragment/json-path-column';
export * from './src/query/ast/fragment/join-fragment';
export * from './src/query/ast/joined-table';
export * from './src/query/ast/select-scalar-expression';
export * from './src/query/ast/set-clause';
export * from './src/query/ast/update-specification';
export * from './src/query/ast/table-reference-expression';
export * from './src/query/ast/json-path-expression';
export * from './src/query/ast/node-part';
export * from './src/query/ast/from-clause';
export * from './src/query/ast/insert-specification';
export * from './src/query/ast/expression/number-literal-expression';
export * from './src/query/ast/expression/raw-binding-expression';
export * from './src/query/ast/expression/expression';
export * from './src/query/ast/expression/binary-expression';
export * from './src/query/ast/expression/condition-factor-expression';
export * from './src/query/ast/expression/common-value-expression';
export * from './src/query/ast/expression/condition-expression';
export * from './src/query/ast/expression/null-predicate-expression';
export * from './src/query/ast/expression/string-literal-expression';
export * from './src/query/ast/expression/raw-expression';
export * from './src/query/ast/expression/exists-predicate-expression';
export * from './src/query/ast/expression/as-expression';
export * from './src/query/ast/expression/function-call-expression';
export * from './src/query/ast/expression/parenthesized-expression';
export * from './src/query/ast/expression/in-predicate-expression';
export * from './src/query/ast/expression/between-predicate-expression';
export * from './src/query/ast/expression/comparison-predicate-expression';
export * from './src/query/ast/expression/condition-term-expression';
export * from './src/query/ast/order-by-clause';
export * from './src/query/ast/from-table';
export * from './src/query/ast/lock-clause';
export * from './src/query/ast/query-specification';
export * from './src/query/ast/identify-variable-declaration';
export * from './src/query/ast/binding-variable';
export * from './src/query/ast/limit-clause';
export * from './src/helper/pluralize';
export * from './src/helper/fn';
export * from './src/helper/obj';
export * from './src/helper/arr';
export * from './src/helper/str';
export * from './src/helper/constructor';
// export * from './src/helper/_pluralize-values';
export * from './src/query-builder/query-builder-resolver-visitor';
export * from './src/query-builder/grammar/query-grammar';
// export * from './src/query-builder/grammar/oracle-query-grammar';
export * from './src/query-builder/grammar/mysql-query-grammar';
export * from './src/query-builder/grammar/postgres-query-grammar';
export * from './src/query-builder/grammar/sqlite-query-grammar';
export * from './src/query-builder/grammar/sqlserver-query-grammar';
export * from './src/query-builder/ast-factory';
export * from './src/query-builder/mixins/union';
export * from './src/query-builder/mixins/having';
export * from './src/query-builder/mixins/where-predicate';
export * from './src/query-builder/mixins/order-by';
export * from './src/query-builder/mixins/where-common';
export * from './src/query-builder/mixins/build-query';
export * from './src/query-builder/mixins/join';
export * from './src/query-builder/mixins/where-date';
export * from './src/query-builder/mixins/group-by';
export * from './src/query-builder/mixins/limit-offset';
export * from './src/query-builder/mixins/aggregate';
export * from './src/query-builder/processor/mysql-processor';
export * from './src/query-builder/processor/postgres-processor';
export * from './src/query-builder/processor/sqlite-processor';
export * from './src/query-builder/processor/sql-server-processor';
export * from './src/query-builder/builder';
export * from './src/query-builder/connection-interface';
export * from './src/query-builder/grammar.interface';
export * from './src/query-builder/processor-interface';
export * from './src/query-builder/processor';
export * from './src/query-builder/visitor/mysql-query-builder-visitor';
export * from './src/query-builder/visitor/sqlserver-query-builder-visitor';
export * from './src/query-builder/visitor/postgres-query-builder-visitor';
export * from './src/query-builder/visitor/query-builder-visitor';
export * from './src/query-builder/visitor/sqlite-query-builder-visitor';
export * from './src/query-builder/forward-ref';
export * from './src/query-builder/query-builder';
export * from './src/query-builder/ast-helper';
export * from './src/database-config';
export * from './src/database-manager';
export * from './src/database-transaction-record';
export * from './src/database-transactions-manager';
export * from './src/manages-transactions';
export * from './src/query-exception';
