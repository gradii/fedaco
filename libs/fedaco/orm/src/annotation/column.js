export class FedacoColumn {
  static isTypeOf(obj) {
    return obj instanceof FedacoColumn
  }
}
