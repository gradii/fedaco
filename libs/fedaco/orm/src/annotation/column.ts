import {
  makeDecorator,
  makePropDecorator,
  TypeDecorator
} from '@gradii/annotation';

export interface ColumnDecorator {

  (obj: Column): TypeDecorator;

  /**
   * See the `Pipe` decorator.
   */
  new(obj: Column): Column;
}

export interface Column {
  name: string;

}

export const Column: ColumnDecorator = makePropDecorator(
  'column',
  (p: Column) => ({ ...p }), undefined, undefined,
  (type: any, meta: Column) => _columnHandle(type, meta)
);

function _columnHandle(type, meta) {

}