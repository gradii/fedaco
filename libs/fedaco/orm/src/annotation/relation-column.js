export class FedacoRelationColumn {
  static isTypeOf(obj) {
    return obj instanceof FedacoRelationColumn;
  }
}
