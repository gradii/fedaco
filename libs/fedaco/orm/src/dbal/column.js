/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license
 */
import { isNumber } from '@gradii/check-type'
import { pascalCase } from '../helper/str'

export class Column {
  constructor(columnName, type, options = {}) {
    this._length = null

    this._precision = 10

    this._scale = 0

    this._unsigned = false

    this._fixed = false

    this._notnull = true

    this._default = null

    this._autoincrement = false

    this._platformOptions = {}

    this._columnDefinition = null

    this._comment = null

    this._customSchemaOptions = {}
    this.setName(columnName)
    this.setType(type)
    this.setOptions(options)
  }
  setName(name) {
    this._name = name
    return this
  }

  setOptions(options) {
    for (const [name, value] of Object.entries(options)) {
      const method = 'set' + pascalCase(name)
      if (method in this) {
        this[method](value)
      } else {
        throw new Error(`The "${name}" column option is not supported`)
      }
    }
    return this
  }

  setType(type) {
    this._type = type
    return this
  }

  setLength(length) {
    if (length !== null) {
      this._length = length
    } else {
      this._length = null
    }
    return this
  }

  setPrecision(precision) {
    if (!isNumber(precision)) {
      precision = 10
    }
    this._precision = +precision
    return this
  }

  setScale(scale) {
    if (!isNumber(scale)) {
      scale = 0
    }
    this._scale = +scale
    return this
  }

  setUnsigned(unsigned) {
    this._unsigned = Boolean(unsigned)
    return this
  }

  setFixed(fixed) {
    this._fixed = Boolean(fixed)
    return this
  }

  setNotnull(notnull) {
    this._notnull = Boolean(notnull)
    return this
  }

  setDefault(_default) {
    this._default = _default
    return this
  }

  setPlatformOptions(platformOptions) {
    this._platformOptions = platformOptions
    return this
  }

  setPlatformOption(name, value) {
    this._platformOptions[name] = value
    return this
  }

  setColumnDefinition(value) {
    this._columnDefinition = value
    return this
  }

  getType() {
    return this._type
  }

  getLength() {
    return this._length
  }

  getPrecision() {
    return this._precision
  }

  getScale() {
    return this._scale
  }

  getUnsigned() {
    return this._unsigned
  }

  getFixed() {
    return this._fixed
  }

  getNotnull() {
    return this._notnull
  }

  getDefault() {
    return this._default
  }

  getPlatformOptions() {
    return this._platformOptions
  }

  hasPlatformOption(name) {
    return this._platformOptions[name] !== undefined
  }

  getPlatformOption(name) {
    return this._platformOptions[name]
  }

  getColumnDefinition() {
    return this._columnDefinition
  }

  getAutoincrement() {
    return this._autoincrement
  }

  setAutoincrement(flag) {
    this._autoincrement = flag
    return this
  }

  setComment(comment) {
    this._comment = comment
    return this
  }

  getComment() {
    return this._comment
  }

  setCustomSchemaOption(name, value) {
    this._customSchemaOptions[name] = value
    return this
  }

  hasCustomSchemaOption(name) {
    return this._customSchemaOptions[name] !== undefined
  }

  getCustomSchemaOption(name) {
    return this._customSchemaOptions[name]
  }

  setCustomSchemaOptions(customSchemaOptions) {
    this._customSchemaOptions = customSchemaOptions
    return this
  }

  getCustomSchemaOptions() {
    return this._customSchemaOptions
  }
  getQuotedName(grammar) {
    return `"${this._name}"`
  }

  toArray() {
    return Object.assign(
      Object.assign(
        {
          name: this._name,
          type: this._type,
          default: this._default,
          notnull: this._notnull,
          length: this._length,
          precision: this._precision,
          scale: this._scale,
          fixed: this._fixed,
          unsigned: this._unsigned,
          autoincrement: this._autoincrement,
          columnDefinition: this._columnDefinition,
          comment: this._comment,
        },
        this._platformOptions
      ),
      this._customSchemaOptions
    )
  }
}
