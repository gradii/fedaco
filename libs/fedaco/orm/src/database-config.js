import { ConnectionFactory } from './connector/connection-factory'
import { DatabaseManager } from './database-manager'
import { NullDispatcher } from './fedaco/mixins/has-events'
import { Model } from './fedaco/model'
export class DatabaseConfig {
  constructor() {
    this.config = {
      database: {
        fetch: 0,
        default: 'default',
        connections: {},
      },
    }

    this.setupManager()
  }

  setupManager() {
    const factory = new ConnectionFactory()
    this.manager = new DatabaseManager(factory)
  }

  static connection(connection = null) {
    return this.instance.getConnection(connection)
  }

  setAsGlobal() {
    this.constructor.instance = this
  }

  static table(table, as = null, connection = null) {
    return this.instance.constructor.connection(connection).table(table, as)
  }

  static schema(connection = null) {
    return this.instance.constructor.connection(connection).getSchemaBuilder()
  }

  getConnection(name = null) {
    return this.manager.connection(name)
  }

  addConnection(config, name = 'default') {
    const connections = this.config.database.connections

    connections[name] = config
    this.config.database.connections = connections
  }

  bootFedaco() {
    Model.setConnectionResolver(this.manager)
    const events = {
      forget(event) {},
      until() {
        return true
      },
      dispatch() {},
    }
    const dispatcher = new NullDispatcher(events)
    if (dispatcher) {
      Model.setEventDispatcher(dispatcher)
    }
  }

  getDatabaseManager() {
    return this.manager
  }
}
