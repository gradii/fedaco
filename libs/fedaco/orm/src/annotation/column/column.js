/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license
 */
import { makePropDecorator } from '@gradii/annotation'
import { _additionalProcessingGetterSetter } from '../additional-processing'
import { FedacoColumn } from '../column'
export const Column = makePropDecorator(
  'Fedaco:Column',
  (p = {}) => Object.assign({}, p),
  FedacoColumn,
  (target, name, decorator) => {
    _additionalProcessingGetterSetter(target, name, decorator)
  }
)
