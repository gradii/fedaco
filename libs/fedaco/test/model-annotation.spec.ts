import { reflector } from '@gradii/annotation';
import { isFunction } from '@gradii/check-type';
import { findLast } from 'ramda';
import { ColumnAnnotation, FedacoColumn } from '../src/annotation/column';
import { FedacoRelationColumn, RelationColumnAnnotation } from '../src/annotation/relation-column';
import { DatabaseConfig } from '../src/database-config';
import { Model } from '../src/fedaco/model';
import { HasMany } from '../src/fedaco/relations/has-many';
import { SchemaBuilder } from '../src/schema/schema-builder';
import { BasicModel, } from './model/basic.model';
import { FedacoBuilderTestModelParentStub } from './model/fedaco-builder-test-model-parent-stub';
import { ArticleModel, HasManyRelationModel, MemberModel } from './model/has-many-relation.model';
import { HasOneRelationModel } from './model/has-one-relation.model';
import { RelationModel } from './model/relation.model';


function _columnInfo(typeOfClazz: any, key: string) {
  const meta = reflector.propMetadata(typeOfClazz);
  return findLast(it => {
    return FedacoRelationColumn.isTypeOf(it);
  }, meta[key]) as ColumnAnnotation;
}

function connection(connectionName = 'default') {
  return Model.getConnectionResolver().connection(connectionName);
}

function schema(connectionName = 'default'): SchemaBuilder {
  return connection(connectionName).getSchemaBuilder();
}

describe('model annotation', () => {
  async function createSchema() {
    await schema('default')
      .create('article_models', table => {
        table.increments('id');
        table.string('member_model_id');
        table.timestamps();
      });
  }

  beforeAll(async () => {
    const db = new DatabaseConfig();
    db.addConnection({
      'driver': 'sqlite',
      // 'database': files.default
      'database': ':memory:'
    });
    db.bootFedaco();
    db.setAsGlobal();
    await createSchema();
  });

  it('instance model', () => {
    const m = new BasicModel();

    expect(m).toBeInstanceOf(Model);
  });

  it('test basic', () => {
    const basic = new BasicModel();

    const metadata = reflector.propMetadata(BasicModel);

    expect(metadata['name'].length).toBe(1);
    expect(metadata['name'][0].field).toBe('name');
  });

  it('test basic model', () => {
    const basic = new BasicModel();
    basic.name  = 'hello';

    const meta = reflector.propMetadata(BasicModel);
    const a    = findLast(it => FedacoColumn.isTypeOf(it), meta['name']);
  });

  it('test relation annoation', () => {
    const relationModel = new RelationModel();

    const meta = reflector.propMetadata(RelationModel);

    expect(meta).toMatchSnapshot('meta');
  });

  it('test relation annotation get should call getAttribute', () => {
    const relationModel = new RelationModel();

    const meta = reflector.propMetadata(RelationModel);

    const spy1 = jest.spyOn(relationModel, 'getAttribute').mockReturnValue('foo');
    const data = relationModel.columnFoo;

    expect(spy1).toBeCalled();
    expect(data).toBe('foo');
  });

  it('test relation annotation get belongs to many', () => {
    const relationModel = new FedacoBuilderTestModelParentStub();

    const meta = reflector.propMetadata(FedacoBuilderTestModelParentStub);

    const data = relationModel.roles;
  });

  it('test relation annotation set relation should throw exception', () => {
    const relationModel = new RelationModel();

    expect(() => {
      relationModel.columnFoo = 'foo';
    }).toThrow('the relation field is readonly');
  });

  it('test annotation hasOne', () => {
    const relationModel = new HasOneRelationModel();

    const meta = reflector.propMetadata(HasOneRelationModel);
    expect(meta).toMatchSnapshot('has one');


  });

  it('test annotation hasMany', () => {
    const relationModel = new HasManyRelationModel();

    const meta = reflector.propMetadata(HasManyRelationModel);
    expect(meta).toMatchSnapshot('has many');

  });

  it('test annotation get hasMany relation', () => {
    const relationModel = new MemberModel();

    const meta = reflector.propMetadata(MemberModel);
    expect(meta).toMatchSnapshot('has many');


    const metaColumnInfo: RelationColumnAnnotation = _columnInfo(MemberModel, 'articles');

    expect(metaColumnInfo).toMatchObject({
      isRelation: true,
      type      : 'HasMany'
    });

    expect(isFunction(metaColumnInfo._getRelation)).toBe(true);

    const hasMany = metaColumnInfo._getRelation(relationModel, 'articles');

    expect(hasMany).toBeInstanceOf(HasMany);
  });

  xit('test annotation get hasMany relation sql', async () => {
    const memberModel = new MemberModel();

    memberModel.id = 101;

    const metaColumnInfo: RelationColumnAnnotation = _columnInfo(MemberModel, 'articles');

    // const spy1 = jest.spyOn(memberModel._connection, 'query');

    const hasMany = metaColumnInfo._getRelation(memberModel, 'articles');

    expect(hasMany).toBeInstanceOf(HasMany);

    const results = await hasMany.getResults();

    expect(results[0] instanceof ArticleModel).toBe(true);

    // tmp test
    expect(results[0]._attributes['sql']).toBe(
      'SELECT * FROM `article_models` WHERE `article_models`.`member_model_id` = ? AND `article_models`.`member_model_id` IS NOT NULL');
    expect(results[0]._attributes['bindings']).toEqual([101]);
  });

  xit('test annotation use direct field access', async () => {
    const memberModel = new MemberModel();

    memberModel.id = 101;

    const results = await memberModel.articles;

    expect(results[0] instanceof ArticleModel).toBe(true);

    // todo fixme tmp test hack code in hasRelationship
    expect(results[0]._attributes['sql']).toBe(
      'SELECT * FROM `article_models` WHERE `article_models`.`member_model_id` = ? AND `article_models`.`member_model_id` IS NOT NULL');
    expect(results[0]._attributes['bindings']).toEqual([101]);
  });

  // it('test annotation ')

});
