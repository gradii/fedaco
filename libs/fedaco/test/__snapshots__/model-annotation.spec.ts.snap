// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`model annotation test annotation get hasMany relation: has many 1`] = `
Object {
  "articles": Array [
    FedacoRelationColumn {
      "_getRelation": [Function],
      "isRelation": true,
      "related": [Function],
      "type": "HasMany",
    },
  ],
  "id": Array [
    FedacoColumn {},
  ],
}
`;

exports[`model annotation test annotation hasMany: has many 1`] = `
Object {
  "columnFoo": Array [
    FedacoRelationColumn {
      "_getRelation": [Function],
      "isRelation": true,
      "related": [Function],
      "type": "HasMany",
    },
  ],
}
`;

exports[`model annotation test annotation hasOne: has one 1`] = `
Object {
  "columnFoo": Array [
    FedacoRelationColumn {
      "_getRelation": [Function],
      "isRelation": true,
      "type": "HasOne",
    },
  ],
}
`;

exports[`model annotation test relation annoation: meta 1`] = `
Object {
  "columnFoo": Array [
    FedacoRelationColumn {
      "_getRelation": [Function],
      "isRelation": true,
      "type": "HasOne",
    },
  ],
}
`;
